File: 2020/2020W46/PreppindataW46.py
import pandas as pd
import re
from datetime import datetime as dt

xlsx = pd.ExcelFile("F:/Data/Incident List.xlsx")

incident = pd.read_excel(xlsx,'Incident List')
incident['Aircraft'] = incident['Incident'].apply(lambda x: re.search('(.*?)(?=\s(at|near)\s)',x).group(1)) 
incident['Date'] = incident['Incident'].apply(lambda x: dt.strptime(
    re.sub('.*([A-Z][a-z]{2})\s(\d+)[a-z]{2}\s(\d{4}).*','\\2-\\1-\\3',x),'%d-%b-%Y').date()) 
incident['Location'] = incident['Incident'].apply(lambda x: re.search('(?<=at\s)(.*?)(?=\son\s)',x).group(1) 
                                                  if re.search('(?<=at\s)(.*?)(?=\son\s)',x) else re.search('(?<=near\s)(.*?)(?=\son\s)',x).group(1))
incident['Incident Description'] = incident['Incident'].apply(lambda x: re.search('(?<=\d{4},\s)(.*$)',x).group(1)) 
finalA = incident[['Date', 'Location', 'Aircraft', 'Incident Description']]

finalB = pd.read_excel(xlsx,'Category') 
finalB['Number of Incidents'] = finalB['Category'].apply(lambda x: len([idt for idt in incident['Incident Description'].tolist() 
                                                                        if re.search(f"({x[0:5].lower()})",idt)]))

File: 2020/2020W46/preppindataw46.R
library(dplyr)
library(readxl)
library(stringr)
library(fuzzyjoin)

Sys.setlocale("LC_ALL","English")
xlsx <- "F:/Data/Incident List.xlsx"
finalA <- read_excel(xlsx, sheet = "Incident List") %>%
  mutate('Date' = as.Date(str_replace_all(`Incident`,'.*([A-Z][a-z]{2})\\s(\\d+)[a-z]{2}\\s(\\d{4}).*','\\2-\\1-\\3'),'%d-%b-%Y'),
         'Location' = if_else(str_detect(`Incident`, '(\\sat\\s.*\\son\\s)'),
                              str_extract(`Incident`,'(?<=at\\s)(.*?)(?=\\son\\s)'), str_extract(`Incident`, '(?<=near\\s)(.*?)(?=\\son\\s)')),
         'Aircraft' = str_extract(`Incident`,'(.*?)(?=\\s(at|near)\\s)'),
         'Incident Description' = str_extract(`Incident`,'(?<=\\d{4},\\s)(.*$)')) %>%
  select(-c('RecordID', 'Incident'))

finalB <- finalA %>% select('Incident Description') %>%
  regex_inner_join(read_excel(xlsx, sheet = "Category") %>% 
                     mutate('Key' = paste0('.*',str_sub(tolower(`Category`),1,5),'.*')), by = c(`Incident Description`='Key')) %>%
  group_by(`Category`) %>%
  summarise('Number of Incidents' = n(), .groups = 'drop')
   
View(finalA)
View(finalB)
